### PR TITLE
fix(core): a non-positive retry budget no longer skips the wrapped call

### PR DESCRIPTION
**Pull request by DABH**
fix(core): a non-positive retry budget no longer skips the wrapped call

## Description

`nat.utils.exception_handlers.automatic_retries` is the retry layer that provider and client plugin packages build on. The provider-authoring guide (`docs/source/extend/custom-components/adding-an-llm-provider.md`) tells plugin authors to add `RetryMixin` to their provider config, and the first-party client packages wire `RetryMixin.num_retries` straight into `patch_with_retry` so every public client method retries on retryable errors (for example `packages/nvidia_nat_langchain/src/nat/plugins/langchain/llm.py`) — the pattern third-party client packages follow.

All four wrapper variants produced by `_retry_decorator` (plain sync, coroutine, sync generator, and async generator) iterate `for attempt in range(retries)`. When the budget is 0 or negative, the range is empty and the wrapped callable is never invoked at all:

- the sync and coroutine wrappers fall through and return `None`,
- the generator wrappers complete without yielding anything.

No exception is raised and nothing is logged, so the failure mode is a silent no-op. Why is a non-positive budget reachable at all? `RetryMixin.num_retries` carried no lower bound, so `num_retries: 0` validated cleanly in YAML and silently skipped every patched call: first-party and third-party provider packages pass the value straight into `patch_with_retry(retries=...)`. A user who sets `num_retries: 0` in a workflow configuration (a natural way to express "do not retry") turns every patched LLM, embedder, or memory client method into a call that never reaches the provider and quietly returns `None` or an empty stream. That is very hard to debug from inside a plugin package, and it is a correctness trap for any plugin author who exposes the retry configuration to end users.

This change closes the trap at both layers.

At the config layer, mandate the documented contract: `RetryMixin.num_retries` is now a positive integer (`gt=0`).

- `num_retries: 0` or a negative value now fails at configuration validation for every mixin-typed config, instead of loading cleanly and silently skipping calls. This is a deliberate behavior change, and it only rejects configurations that were already broken: "do not retry" remains expressible via `do_auto_retry: false`, and `num_retries: 1` means one total attempt (the initial call, zero retries).
- A new data-model test pins the constraint — `num_retries=0` and `num_retries=-1` are rejected with a `ValidationError` and `num_retries=1` is accepted — both on the mixin itself and on a provider config that inherits it alongside other mixins.

At the wrapper layer, keep the guarantee that the wrapped callable executes at least once (at least one attempt, zero retries):

- Normalize the budget once in `_retry_decorator` (`total_attempts = max(1, retries)`) and use it consistently in all four wrapper variants, both for the attempt loop and for the last-attempt re-raise check.
- Log the normalization once at `DEBUG` level. It happens in the decorator factory, which runs once per `patch_with_retry` call rather than per patched method or per invocation, so the adjustment is observable without adding log noise.
- Document the semantics in the module docstring and in the `_retry_decorator` and `patch_with_retry` docstrings: the budget counts total attempts (the initial call plus any retries), and values below 1 behave as 1.
- Behavior for budgets >= 1 is unchanged; the existing tests that pin attempt counts on persistent failure pass unmodified, and a new boundary case pins budget 1 (exactly one attempt on success and on failure), the point where the normalization becomes a no-op.
- Add tests covering budgets of 0, -1, and the boundary budget of 1 for all four wrapper variants (real result returned, exactly one invocation), failure propagation after a single attempt, and the `patch_with_retry` public path. The eleven non-positive-budget cases fail against the previous implementation; the budget-1 cases pass against both implementations.

Keeping the wrapper clamp alongside the config constraint is deliberate: the field constraint only protects mixin-typed configs. A config class that redeclares `num_retries` replaces the constrained field (an existing in-tree test exercises exactly that shape), and `patch_with_retry` is a public API that can be called directly with any integer. The clamp keeps "at least one attempt" an invariant of the retry layer itself rather than a property of a single validation path.

No tracking issue exists for this yet; happy to file one if the team prefers.

## Testing

- `uv run --project packages/nvidia_nat_core -- pytest packages/nvidia_nat_core/tests/nat/utils/test_retry_wrapper.py -q` — 46 passed (30 pre-existing tests unchanged, 16 new).
- Verified the new tests capture the bug: with the previous implementation restored, the same command reports 11 failed, 35 passed — exactly the non-positive-budget cases fail, while the budget-1 boundary cases pass on both implementations, as intended.
- `uv run pre-commit run yapf --files packages/nvidia_nat_core/src/nat/utils/exception_handlers/automatic_retries.py packages/nvidia_nat_core/tests/nat/utils/test_retry_wrapper.py` — Passed.
- `uv run pre-commit run ruff-check --files packages/nvidia_nat_core/src/nat/utils/exception_handlers/automatic_retries.py packages/nvidia_nat_core/tests/nat/utils/test_retry_wrapper.py` — Passed.
- `uv run --no-sync pytest packages/nvidia_nat_core/tests/nat/data_models/test_retry_mixin.py packages/nvidia_nat_core/tests/nat/utils/test_retry_wrapper.py -q` — 54 passed (8 new constraint tests; the 46 wrapper tests unchanged).
- `uv run --no-sync pytest packages/nvidia_nat_autogen/tests/test_llm_autogen.py -q` — 52 passed, including the pre-existing edge-case test whose config subclass redeclares `num_retries: int = 0`: the redeclaration replaces the constrained field, which is exactly the escape the wrapper clamp still covers.
- `uv run pre-commit run yapf --files packages/nvidia_nat_core/src/nat/data_models/retry_mixin.py packages/nvidia_nat_core/tests/nat/data_models/test_retry_mixin.py` — Passed; same files with `ruff-check` — Passed.
- `uv run python ci/scripts/copyright.py --verify-apache-v2` — passed.

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/NeMo-Agent-Toolkit/blob/develop/docs/source/resources/contributing/index.md).
- We require that all contributors "sign-off" on their commits. This certifies that the contribution is your original work, or you have rights to submit it under the same license, or a compatible license.
  - Any contribution which contains commits that are not Signed-Off will not be accepted.
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Retry settings below one attempt now execute exactly once instead of being skipped.
  * Retry behavior is consistent across synchronous, asynchronous, and generator-based operations.
  * Initial failures are correctly surfaced when no additional retries are available.
  * Invalid non-positive retry settings are rejected in configuration.

* **Documentation**
  * Updated retry guidance to clarify that retry values represent the total attempt budget.

### CODE REVIEW

**Comment by chuenchen309** [comment]
Checked this against `develop`. The bug is real, the fix is right, and the reachability argument is stronger than the description makes it — worth adding, because it's the part a reviewer will want before merging.

`automatic_retries.py` imports only the standard library, so I loaded the file straight off `develop` and drove all four wrapper shapes rather than transcribing anything:

```
  --- retries=3  (control) ---
    sync      -> 'sync-result'      generator -> ['gen-item']
    coroutine -> 'coro-result'      async gen -> ['agen-item']
    wrapped callable actually invoked: ['sync', 'coro', 'gen', 'agen']

  --- retries=0 ---
    sync      -> None               generator -> []
    coroutine -> None               async gen -> []
    wrapped callable actually invoked: NEVER
```

`-1` behaves the same. So it isn't only that the call is skipped — the sync and coroutine wrappers hand the caller `None` and the two generator wrappers hand back an empty sequence, both of which look like a legitimate result rather than a failure.

**Why a non-positive budget is reachable.** `RetryMixin.num_retries` is `Field(default=5, ...)` with no bound — no `ge`, no `gt`:

```
num_retries= 5  -> ACCEPTED        num_retries= 0  -> ACCEPTED
num_retries= 1  -> ACCEPTED        num_retries=-1  -> ACCEPTED
```

So `num_retries: 0` in a workflow YAML validates cleanly, and it's the obvious way for someone to write "don't retry this provider" — `do_auto_retry: false` is the intended switch, but nothing points you there or rejects the other spelling. The user gets an LLM client whose methods silently return `None`.

**The fix doesn't change the normal path.** Counting invocations of an always-failing callable, before and after the patch:

```
              develop        with #2146
retries=1     1 time(s)      1 time(s)
retries=3     3 time(s)      3 time(s)
retries=5     5 time(s)      5 time(s)
```

Identical, which confirms the docstring addition ("counts *total attempts*: the initial call plus any retries") is documenting the existing semantics rather than redefining them — worth stating explicitly, since "normalize to at least 1 attempt" could otherwise read like a behaviour change to the budget.

One detail worth calling out because it's easy to miss when reviewing: changing `attempt == retries - 1` to `attempt == total_attempts - 1` isn't cosmetic. Left alone with `retries=0`, that comparison is against `-1` and never matches, so the last-attempt branch would stop raising. Both had to move together, and they did in all four wrappers.

Nothing blocking from me. The only thing I'd consider is whether `RetryMixin.num_retries` should also carry `ge=1`, so the config surface rejects the value instead of the decorator quietly repairing it — but that's a separate call about config validation, and this PR is the right fix for the layer it's touching.

*AI-assisted review; the outputs above come from executing the module as fetched from `develop`, and from applying this PR's change to it, not from reading. I have no stake in this PR.*

**Review by willkill07 — CHANGES_REQUESTED** [review]
Rather than ensure at least one retry, why not mandate the contract of `RetryMixin.num_retries` be a positive integer? Pydantic allows you to specify concrete bound requirements on values:

- field attribute condition: `Field(gt=0)`
- importing `PositiveInteger` from pydantic
- custom definition: `PositiveInteger = Annotated[int, Field(gt=0)]`

**Comment by chuenchen309** [comment]
On the `Field(gt=0)` suggestion — I measured what it would do on `develop`, because the answer surprised me in both directions.

**It does not bind the shape this repo actually uses in tests.** Pydantic replaces the parent `FieldInfo` when a subclass redeclares the field, so a subclass that writes `num_retries: int = 0` escapes the constraint entirely (pydantic 2.13.4, inside the `pydantic~=2.11` pin):

```
With gt=0 on the mixin:
  mixin explicit 0                  -> ValidationError: Input should be greater than 0
  subclass, no redeclaration, 0     -> ValidationError: Input should be greater than 0
  subclass redeclares `int = 0`     -> num_retries=0     <-
  subclass redeclares, explicit -1  -> num_retries=-1    <-
```

`test_llm_autogen.py::TestErrorHandling::test_patch_with_invalid_retry_config` is exactly that shape — it declares `class InvalidRetryConfig(LLMBaseConfig, RetryMixin)` with `num_retries: int = 0` and asserts it "should not raise, even with edge case values". So that test would keep passing with a zero budget, and the wrapper it exercises would still be the thing deciding what a zero means.

**It does bind every provider config.** `RetryMixin` is inherited by 12 production config classes in `nvidia_nat_core`: seven LLM configs (`OpenAIModelConfig`, `NIMModelConfig`, `AzureOpenAIModelConfig`, `AWSBedrockModelConfig`, `LiteLlmModelConfig`, `OCIModelConfig`, `HuggingFaceInferenceLLMConfig`), four embedder configs (`OpenAIEmbedderModelConfig`, `NIMEmbedderModelConfig`, `AzureOpenAIEmbedderModelConfig`, `HuggingFaceEmbedderConfig`), and `EvaluatorLLMConfig`. `grep -rn 'num_retries\s*:' --include='*.py' . | grep -v /tests/` shows none of them redeclares it, so `gt=0` on the mixin makes `num_retries: 0` a config validation error for every LLM and embedder in the toolkit.

That matters because `num_retries` has a second consumer that has nothing to do with this wrapper: eleven plugin packages forward it verbatim into third-party clients as `retries=llm_config.num_retries` (langchain, llama_index, agno, crewai, strands, semantic_kernel, autogen, memmachine, mem0ai, zep_cloud, ragas). For those SDKs zero is a meaningful setting — "call once, do not retry" — and it is currently the only way to express it.

So the two changes are not alternatives to each other in the way the comment reads. The constraint is leaky against the redeclaring subclasses already in-tree, and it removes a value that is legal on the paths that bypass this wrapper; the wrapper-level guard covers the case the PR is about regardless of where the number came from. Doing both is coherent — the constraint would just be a config-surface decision about whether `0` should stay expressible, which is yours to make rather than mine.

*AI-assisted; the table is output from running the two class shapes under the pinned pydantic, and the inheritance counts are from greps against `develop` at `8848b59`.*

**Comment by DABH** [comment]
Good call — adopted. The branch now adds `gt=0` to `RetryMixin.num_retries`, so `num_retries: 0` (or negative) fails at config validation with a clear pydantic error instead of loading and silently no-opping. Flagging the deliberate behavior change that comes with it: configs that previously set `num_retries: 0` were already silently broken (every patched call skipped), and "don't retry" stays expressible via `do_auto_retry: false` — the intended switch — or `num_retries: 1`, which under the wrapper's semantics means one total attempt.

I'd like to keep the wrapper-level guard alongside the constraint, for two reasons that are quick to verify:

1. Pydantic constraints don't survive subclass redeclaration: a config class that redeclares `num_retries: int = 0` replaces the field's metadata and accepts 0 (and even -1) despite the mixin's bound — and that exact shape exists in-tree (`test_llm_autogen.py`'s `InvalidRetryConfig`). Happy to paste the 4-line repro.
2. `patch_with_retry` is public API that provider packages call directly (per `adding-an-llm-provider.md`), so the value doesn't always arrive through the mixin. The wrapper is the one layer every path crosses, which is where an "executes at least once" guarantee has to live.

One clarification to my own framing, since your comment said "ensure at least one retry": the guarantee is at least one *attempt* (i.e., zero retries) — I've updated the PR description to say that precisely.

File: packages/nvidia_nat_core/src/nat/data_models/retry_mixin.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ class RetryMixin(BaseModel):
                                 " that fail with a retryable error.",
                                 exclude=True)
     num_retries: int = Field(default=5,
+                             gt=0,
                              description="Number of times to retry a method call that fails"
                              " with a retryable error.",
```
[the remark this answers is not in the document]
Comment (DABH) [reply packages/nvidia_nat_core/src/nat/data_models/retry_mixin.py]: Applied in 198602a5 — the description now states the total-attempt contract (including the initial call), matching the wrapper semantics discussed above.

File: packages/nvidia_nat_core/tests/nat/data_models/test_retry_mixin.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import pytest
+from pydantic import ValidationError
+
+from nat.data_models.retry_mixin import RetryMixin
+from nat.llm.openai_llm import OpenAIModelConfig
+
+
+class TestRetryMixin:
+    """Tests for RetryMixin field validation."""
```
[the remark this answers is not in the document]
Comment (DABH) [reply packages/nvidia_nat_core/tests/nat/data_models/test_retry_mixin.py]: Done in 198602a5 — both docstrings now quote `RetryMixin`.

**Review by willkill07 — APPROVED** [review]

**Comment by willkill07** [comment]
/ok to test [198602a](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/2146/commits/198602a5540110e027dba8e4564cb1482c28de5c)
